File: 345.反转字符串中的元音字母.py
#
# @lc app=leetcode.cn id=345 lang=python3
#
# [345] 反转字符串中的元音字母
#

# @lc code=start


class Solution:
    def reverseVowels(self, s: str) -> str:
        s = list(s)
        start, end = 0, len(s) - 1
        yuan = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'}
        while start < end:
            while start < end and s[start] not in yuan:
                start += 1
            while start < end and s[end] not in yuan:
                end -= 1
            s[start], s[end] = s[end], s[start]
            start += 1
            end -= 1
        return "".join(s)

# @lc code=end

File: 345.反转字符串中的元音字母.java
/*
 * @lc app=leetcode.cn id=345 lang=java
 *
 * [345] 反转字符串中的元音字母
 */

// @lc code=start
class Solution {
    public String reverseVowels(String s) {
        char[] arr = s.toCharArray();
        int start = 0;
        int end = arr.length-1;
        char tmp;
        while (start < end) {
            while (start < end && !isYuan(arr[start]))
                start++;
            while (start < end && !isYuan(arr[end]))
                end--;
            tmp = arr[start];
            arr[start] = arr[end];
            arr[end] = tmp;
            start++;
            end--;
        }
        // StringBuilder sb = new StringBuilder
        return new String(arr);
    }
    private boolean isYuan(char ch) {
        switch (ch) {
            case 'a':
            case 'e':
            case 'i':
            case 'o':
            case 'u':
            case 'A':
            case 'E':
            case 'I':
            case 'O':
            case 'U': return true;
            default: return false;
        }
    }
}
// @lc code=end

